### ISSUE TITLE
caused "not a directory"

### ISSUE THREAD

**Issue by bupy7** [docker/compose#3978]
I cannot mount the nginx config file:

```
version: '2'
services:
    web:
        image: nginx:latest
        ports:
            - "8080:80"
        volumes:
            - ./nginx/nginx.conf:/etc/nginx/conf.d/default.conf
```

Error:
```
$ docker-compose up
Creating network "docker_default" with the default driver
Creating docker_web_1

ERROR: for web  Cannot start service web: oci runtime error: rootfs_linux.go:53: mounting "/mnt/sda1/var/lib/docker/aufs/mnt/9ef61a113b176b7175ed0f870828b29664300ac9a5dea57fe1987dfa308633be/etc/nginx/conf.d/default.conf" to rootfs "/mnt/sda1/var/lib/docker/aufs/mnt/9ef61a113b176b7175ed0f870828b29664300ac9a5dea57fe1987dfa308633be" caused "not a directory"
ERROR: Encountered errors while bringing up the project.
```

Via docker:

```
$ docker run --rm -it -v $(pwd)/nginx/nginx.conf:/etc/nginx/conf.d/default.conf nginx:latest
docker: Error response from daemon: oci runtime error: rootfs_linux.go:53: mounting "/mnt/sda1/var/lib/docker/aufs/mnt/1d488c9f67c62b352ba91ca537bd72666b1b635d88e112a3ad4bf3cc63600a8a/etc/nginx/conf.d/default.conf" to rootfs "/mnt/sda1/var/lib/docker/aufs/mnt/1d488c9f67c62b352ba91ca537bd72666b1b635d88e112a3ad4bf3cc63600a8a" caused "not a directory".
```

OS: OSX 10.9.5
Docker: 
```
$ docker --version
Docker version 1.12.0, build 8eab29e

$ docker-machine --version
docker-machine version 0.8.0, build b85aac1

$ docker-compose --version
docker-compose version 1.8.0, build f3628c7
```

**Comment by miguelcudaihl** [comment]
I encountered the same problem.  Ensure that '/etc/nginx/conf.d/default.conf' exists on your host.

**Comment by bupy7** [comment]
@miguelcudaihl `/etc/nginx/conf.d/default.conf` exists on my host:

docker-compose.yml:

```
version: "2"
services:
    web:
        image: nginx:latest
        ports:
            - "8080:80"
        command: ls -l /etc/nginx/conf.d
```

Out:
```
$ docker-compose up
Recreating docker_web_1
Attaching to docker_web_1
web_1  | total 4
web_1  | -rw-r--r-- 1 root root 1097 Sep 13 16:18 default.conf
docker_web_1 exited with code 0
```

**Comment by shin-** [comment]
Hi!

First off, if `docker run` gives you the exact same error message, chances are it's not actually a Compose error and thus should not be reported here.

Now to answer your question, you can only mount directories using volume bindings - not single files.

HTH!

**Comment by jacekszubert** [comment]
> you can only mount directories using volume bindings - not single files
Due to limitations of docker or compose? According to the [https://docs.docker.com/engine/tutorials/dockervolumes/#/mount-a-host-file-as-a-data-volume](docker docs) it is possible. I can't find this information in compose docs, but it was always possible - has something changed lately here? Can't find information in the changelog neither.

**Comment by shin-** [comment]
Oops - I don't know what I was thinking. You're right, that is possible.

My other point still stands however - since the behavior is identical with `docker run`, this is a bug in engine and should be reported there.

**Comment by settermjd** [comment]
I'm pretty new to Docker (at least deploying with it), but I can both ls and cat default.conf, yet I get this error.

**Comment by shin-** [comment]
@settermjd https://github.com/moby/moby/issues/27010#issuecomment-250437911

**Comment by serafss2** [comment]
I had the same issue because I did not create the file on the host before running the `docker-compose` command, so the `docker-compose` command automatically created a folder which makes sense.

When I deleted the folder and created the file at the same location, that error popped up, I'm suspecting there's some cache that tagged that `volume` as a folder.

To solve the issue I removed the volume part from the `docker-compose.yml`, ran the `docker-compose` command again, then I put back the `volume` part back to the file and it worked.

#### TL;DR
1. Remove the Volume part from the `docker-compose.yaml` file
2. Run your `docker-compose up -d` command
3. Put back the Volume part into the `docker-compose.yaml` file
4. Run your `docker-compose up -d`command again

**Comment by jmadelaine** [comment]
I solved this by removing the volumes from docker. Using `docker volume ls` will show all volumes, you can then remove the specific volume using:

```
docker volume rm <volume_name_here>
```

Rebuilding the container will create a new copy of the volume.

**Comment by matrumz** [comment]
@serafdev 

This is my exact scenario, but your fix is not working for me. Neither is @jmadelaine 's. Any other ideas?

**Comment by serafdev** [comment]
@matrumz docker-compose has changed a lot since this issue was opened, are you able to add details on how we can reproduce your issue/include some logs, a snippet of your yaml file, etc?

**Comment by matrumz** [comment]
Docker Client: Version: 20.10.7
Docker Engine Version: 20.10.11
containerd Version: 1.4.12
runc Version: 1.0.2
docker-init Version: 0.19.0
Docker Compose Version: 2.2.2
OS: WSL2 - Ubuntu 20.04

docker-compose.yaml:
```yaml
version: "3.9"
 
common: &common
  build:
    context: .
    dockerfile: Dockerfile
  volumes:
    - ${HOME}/.gitconfig:/root/.gitconfig

services:
  "something":
    <<: *common
```

- Used a compose file similar to this to build and launch a container - hadn't realized that a `.gitconfig` file didn't yet exist on host machine
- Container started - `/root/.gitconfig` was created as a directory (as expected if the host file didn't exist)
- did `compose down` to take down container
- deleted `.gitconfig` directory created on the host
- created the `.gitconfig` file on the host
- did `compose up -d --force-recreate` to bring new container up
  - resulted in error: 
```sh
⠋ Container <name-redacted>  Creating         0.0s
Error response from daemon: not a directory
```

Tagging @brian-toner to help with any further data collection.

**Comment by serafdev** [comment]
Is `~/.gitconfig` a directory on the host after your first `compose up`?

**Comment by serafdev** [comment]
Do you have a `.gitignore` originally in the container @matrumz? Or does it create it the first time you run it? Can you try creating the `.gitignore` as a file on the host machine before running your container?

For the sake of testing, name it `~/.gitignore.test` and change that in your config above (`${HOME}/.gitignore.test:...`)

**Comment by matrumz** [comment]
No, and I'm not sure what the relevance is: could you explain?

The image we are creating a container from has no repos in it. When the container is created, we clone a git repo into the container to do development. In order to keep a common git experience, we mount the host `.gitconfig`, _not `.gitignore`_, into the container. 

When one of our developers attempted the compose, they hadn't had a `.gitconfig` in their home dir at the time because they hadn't used their WSL instance for git work yet. Hence the issue we're running into now.

**Comment by brian-toner** [comment]
So, I did create the .gitignore, but it did not work.  So, some of the trouble here is that I have one WSL that will run the docker image and a second that will not.  The second WSL I actually setup the .gitconfig before starting the docker image.  On the first WSL I did not.  So I can run the docker image, just not on the first WSL.  I haven't had a chance to troubleshoot any deeper than that.

**Comment by brian-toner** [comment]
So I do have an update on this.  After some windows updates and a couple of restarts, this problem seems to have resolved itself.  There may have been some kind of temp file that was mucking things up.

**Comment by TBG-FR** [comment]
Encountering the same issue here... Any updates ?

**Comment by bupy7** [comment]
@TBG-FR, I doesn't remember how I fixed it, but can suggest that file/directory should not exists before mounting. For my situation this was file `/etc/nginx/conf.d/default.conf` that already exists. Try drop your target file/directory before mounting or change target path.

**Comment by kl3m3n** [comment]
Had the same issue as OP. I run "docker compose down" and then "docker compose up -d" worked just as expected.
